release/nightly_tests/dataset/benchmark.py

```diff
@@ -50,32 +43,82 @@ def __init__(self, name):
         self.result = {}
         print(f"Running benchmark: {name}")
 
-    def run(self, name: str, fn: Callable[..., Dataset], **fn_run_args):
+    def run_materialize_ds(
+        self,
+        name: str,
+        fn: Callable[..., Dataset],
+        *fn_args,
+        **fn_run_args,
+    ):
+        """Run a benchmark on materializing a Ray Dataset. ``fn`` is expected to
+        return the Dataset which is to be materialized. Runtime and throughput
+        are automatically calculated and reported."""
+
         gc.collect()
 
         print(f"Running case: {name}")
         start_time = time.perf_counter()
-        output_ds = fn(**fn_run_args)
+        output_ds = fn(*fn_args, **fn_run_args)
         output_ds.materialize()
         duration = time.perf_counter() - start_time
 
         # TODO(chengsu): Record more metrics based on dataset stats.
-        self.result[name] = {BenchmarkMetric.RUNTIME.value: duration}
+        self.result[name] = {
+            BenchmarkMetric.RUNTIME.value: duration,
+            BenchmarkMetric.THROUGHPUT.value: output_ds.count() / duration,
+        }
+        print(f"Result of case {name}: {self.result[name]}")
+
+    def run_iterate_ds(
+        self,
+        name: str,
+        ds_iterator: Iterator[Any],
+        batch_size: int,
+    ):
+        """Run a benchmark iterating over a dataset (not limited to Ray Data,
+        e.g. could be an iterator over Torch dataloader). ``ds_iterator`` is expected
+        to be an iterator over the dataset, yielding batches of the given size.
+        Runtime and throughput are automatically calculated and reported."""
+
+        gc.collect()
+
+        print(f"Running case: {name}")
+        start_time = time.perf_counter()
+        record_count = 0
+        for batch in ds_iterator:
+            # Note, we are using `batch_size` as an approximate way to get the total
+            # record count, because each batch type has its own method of getting
+            # batch length, also may depend on the underlying data structure, etc.
+            record_count += batch_size
```

Comment:
To avoid this approximation, any reason we can't call `Dataset.count()` like we do for `run_materialize_ds`?

Reply:
this method is intended to work for not only ray datasets, but also other dataset iterators/loaders like torch, mosaicml, etc. so it's a bit difficult to generalize the logic for getting the number of rows, especially without knowing the underlying structure of the data

Reply:
I guess we only need to support a known number of iterators, right? Can we add a parameter for the iterable type? so here we can get the actual batch size based on the type.

Reply:
Added some logic to automatically calculate the size of each batch depending on the underlying type.